File: JobSearchCodeExamples/JobSearchCodeExamples.VB/TestTaker/EmployeeCoverage.vb
Public Class EmployeeCoverage

    ''' <summary>
    ''' Maximums the coverage.
    ''' </summary>
    ''' <paramname="employees">The employees.</param>
    ''' <returns></returns>
    Public Shared Function MaxCoverage(
            ByVal employees As String()) As Integer

        Dim combined As Integer() = New Integer(employees.Length - 1) {}

        For i As Integer = 0 To employees.Length - 1
            Dim years = employees(i).Split("-")
            Dim startDate = Integer.Parse(years(0))
            Dim endDate = Integer.Parse(years(1))
            For j As Integer = i + 1 To employees.Length - 1
                Dim years2 = employees(j).Split("-")
                Dim start = Integer.Parse(years2(0))
                Dim [end] = Integer.Parse(years2(1))
                If startDate <= start AndAlso endDate >= start OrElse
                   startDate <= [end] AndAlso endDate >= [end] OrElse
                   startDate <= start AndAlso endDate >= [end] OrElse
                   startDate >= start AndAlso endDate <= [end] Then
                    combined(i) += If(combined(i) = 0, 2, 1)
                    combined(j) += If(combined(j) = 0, 2, 1)
                End If
            Next
        Next
        Dim maxValue = 0
        For i As Integer = 0 To employees.Length - 1
            If combined(i) > maxValue Then maxValue = combined(i)
        Next

        Return maxValue

    End Function

End Class

File: JobSearchCodeExamples/JobSearchCodeExamples.cpp/EmployeeCoverage.cpp
#include "EmployeeCoverage.h"

int EmployeeCoverage::MaxCoverage(array<String^>^ employees)
{
	array<int>^ combined = gcnew array<int>(employees->Length);
	for (int i = 0; i < employees->Length; i++)
	{
		array<String^>^ years = employees[i]->Split('-');
		int startDate = Int32::Parse(years[0]);
		int endDate = Int32::Parse(years[1]);
		for (int j = i + 1; j < employees->Length; j++)
		{
			array<String^>^ years2 = employees[j]->Split('-');
			int start = Int32::Parse(years2[0]);
			int end = Int32::Parse(years2[1]);
			if (startDate <= start && endDate >= start ||
				startDate <= end && endDate >= end ||
				startDate <= start && endDate >= end ||
				startDate >= start && endDate <= end)
			{
				combined[i] += combined[i] == 0 ? 2 : 1;
				combined[j] += combined[j] == 0 ? 2 : 1;
			}
		}
	}

	int maxValue = 0;
	for (int i = 0; i < employees->Length; i++)
		if (combined[i] > maxValue) maxValue = combined[i];

	return maxValue;
}
